caffe2/serialize/intermediate_model.cc

```diff
@@ -0,0 +1,334 @@
+  protoVersion_ = model_def->proto_version();
+  AT_ASSERTM(model_def->has_main_module(), "main_module is required for ModelDef.");
+  mainModule_.update(model_def->mutable_main_module(), id_data, mode);
+}
+
+void IntermediateModel::dump(torch::ModelDef* model_def) {
+  model_def->set_name(name_);
+  model_def->set_producer_name(producerName_);
+  model_def->set_producer_version(producerVersion_);
+  model_def->set_proto_version(protoVersion_);
+  mainModule_.dump(model_def->mutable_main_module());
+}
+
+void serializeIntermediateModel(IntermediateModel* imodel,
+    torch::jit::PyTorchFileWriter* writer) {
+  std::unordered_map<void*, uint64_t> data_id;
+  std::stack<IntermediateModule*> imodules;
+  imodules.push(imodel->mutableMainModule());
+  while (!imodules.empty()) {
+    IntermediateModule* m = imodules.top();
+    imodules.pop();
+    auto* params = m->mutableParameters();
+    for (int i = 0; i < params->size(); ++i) {
+      std::shared_ptr<SharedData> dataptr = params->at(i).mutableTensor()->data();
+      void* data = dataptr->dataPtr.get();
+      size_t size = dataptr->size;
+      auto it = data_id.find(data);
+      if (it != data_id.end()) {
+        dataptr->recordId = it->second;
+      } else {
+        uint64_t id = writer->writeRecord(data, size);
+        dataptr->recordId = id;
+        data_id[data] = id;
+      }
+    }
+
+    auto* subms = m->mutableSubmodules();
+    for (int i = 0; i < subms->size(); ++i) {
+      imodules.push(&subms->at(i));
+    }
+  }
+
+  torch::ModelDef model_def;
+  imodel->dump(&model_def);
+  std::string buffer;
+  model_def.SerializeToString(&buffer);
+  writer->writeRecord(buffer.c_str(), buffer.size());
+}
+
+void serializeIntermediateModel(IntermediateModel* imodel, const std::string& filename) {
+  torch::jit::PyTorchFileWriter writer(filename);
+  serializeIntermediateModel(imodel, &writer);
+  writer.writeEndOfFile();
+}
+
+void deserializeIntermediateModel(IntermediateModel* imodel,
+    torch::jit::PyTorchFileReader* reader, DeserializeMode mode) {
+  std::unordered_map<uint64_t, std::shared_ptr<SharedData>> id_data;
+  if (mode == DeserializeMode::LAZY) {
+    // only load model meta data, no tensor data
+    at::DataPtr data_ptr;
+    size_t data_size;
+    std::tie(data_ptr, data_size) = reader->getLastRecord();
+    torch::ModelDef model_def = torch::ModelDef();
+    model_def.ParsePartialFromArray(data_ptr.get(), data_size);
+    imodel->update(&model_def, &id_data, mode);
+    return;
+  }
+  AT_ASSERT(mode == DeserializeMode::EAGER);
+  while (reader->hasNextRecord()) {
+    at::DataPtr data_ptr;
+    size_t data_key;
+    size_t data_size;
+    std::tie(data_ptr, data_key, data_size) = reader->getNextRecord();
+    if (!reader->hasNextRecord()) {
+      // the last record is model data (ModelDef)
+      torch::ModelDef model_def = torch::ModelDef();
+      model_def.ParsePartialFromArray(data_ptr.get(), data_size);
```

Comment:
I think you need to check the result value. And why Partial version?

Reply:
Agree ParseFromArray is more restricted, we should use it and ASSERT on the results.